proposed/before_bundle_build_hook.md

```diff
@@ -0,0 +1,53 @@
+
+Requiring that constituent files are signed prior to building a singlefile app could be highly inconvenient to existing build pipelines.
+It is particularly problematic for a file that exists only in a transient form before its inclusion in a single file bundle. This includes the app main binary itself (as in `hello.dll` for a hello app) and composite r2r image, when composite r2r is used. The build does not currently present a documented opportunity for the user to observe and modify files before inclusion into the bundle.
+
+In addition to signing, similar difficulties may affect other scenarios that rely on post-build modification of build artefacts. Some common modifications include obfuscation and instrumentation.
+
+## Current state (as of .NET 6)
+
+The problem of signing applies primarily to files that are extracted to the file system as a part of execution. For the assemblies that are not extracted, this problem is generally not a concern. 
+The transition to using `superhost` as the host technology for singlefile apps on all platforms in 6.0 means the need to extract files is greatly reduced. Native binaries that belong to the .NET platform are no longer extracted and in fact cannot be extracted, since they are monolithically linked with the app. 
+
+The files that still can be extracted:
+-	Native dependencies included via a nuget references. (this is, technically, the only truly supported scenario to include a native dependency)
+It could be expected that such files are signed prior to the build.
+
+-	Dependencies that are forced to be extracted by the user.  
+There are several possible reasons for the user to force extraction of all files and there is a way to force extraction of all dependencies, including managed assemblies that otherwise could run in-place.
+Note that these files include the app dll itself as well as `.r2r.dll` in a case of composite publishing. Both would not present themselves to the user in the process of building.
+This is the case where extraction is a very observable feature and it is plausible the extraction as a feature will be carried forward in the future versions.
+I believe this is the case that causes the most inconveniences.
+
+The abscense of documented solutions in this area forces users to implement solutions that rely on internal implementation details.  
+For example: [`CollectBundleFilesToSign` target](https://github.com/dotnet/diagnostics/blob/18a8e986ddf28ee3c950b5acf00abc652d87be74/src/Tools/Directory.Build.targets).  
+
+## Proposal for .NET 6
+
+Provide a build-time hook to implement signing logic. The use pattern is expected to be similar to `CollectBundleFilesToSign`, but without relying on undocumented behavior.  
+We will provide:
+-	A target `BeforeBundle` that will be called between `_ComputeFilesToBundle` and `GenerateBundle`
```

Comment:
Looking at the current PR for implementing this - I was under the assumption that the `BeforeBundle` would be target defined by the application - it would be similar to what `BeforeBuild` does - meaning that the right properties would be already computed before it is triggered.
I don't think this design is wrong - it's just that the PR doesn't match it (or I don't understand how MSBuild works - which is very possible).

Reply:
MSBuild scripting is not something that I do every day either :-). I am not sure how a user can make their target be called directly, there is probably a way. It is however pretty simple to insert a target into the dependence order between existing targets via `BeforeTargets`, `DependsOnTargets`, so I always assumed that defining documented points between which user inserts their stuff _is_ the way to provide a stable extensibility point in MSBuild.

This is also the way the existing signing support works, although it uses undocumented `_ComputeFilesToBundle`/`_FilesToBundle`

Reply:
Great - we should just include a sample of how this will be done in this doc then.

Reply:
CallTarget is how you do it. But that's something we actively discourage except for some really specific cases. Indeed, the Before + Depends is probably the best way.